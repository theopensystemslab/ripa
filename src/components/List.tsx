import Box from "@material-ui/core/Box";
import IconButton from "@material-ui/core/IconButton";
import InputBase from "@material-ui/core/InputBase";
import MenuItem from "@material-ui/core/MenuItem";
import Select from "@material-ui/core/Select";
import { makeStyles } from "@material-ui/core/styles";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableContainer from "@material-ui/core/TableContainer";
import TableFooter from "@material-ui/core/TableFooter";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import CloseIcon from "@material-ui/icons/Close";
import * as React from "react";

const useStyles = makeStyles(theme => ({
  table: {
    width: "auto",
    tableLayout: "auto"
  },
  rowCount: {
    width: "2rem",
    textAlign: "center",
    borderLeft: "0 !important",
    "& + *": {
      borderLeft: "0 !important"
    }
  },
  rowFocus: {
    backgroundColor: theme.palette.grey[100]
  },
  deleteRow: {
    width: "2rem",
    textAlign: "center",
    padding: 0,
    borderLeft: "0 !important"
  },
  inputCell: {
    padding: 0
  },
  input: {
    padding: theme.spacing(1.5),
    height: "100%",
    "& input": {
      padding: 0
    }
  },
  selectInput: {
    width: "100%",
    border: "0 !important",
    boxShadow: "none !important"
  },
  select: {
    padding: theme.spacing(1.5, 3, 1.5, 1.5),
    border: 0,
    height: "100%",
    width: "100%",
    display: "block"
  },
  selectMenu: {
    "& > ul": {
      padding: 0
    }
  }
}));

const InputCell = ({ colSpan = 1, ...props }) => {
  const classes = useStyles();
  return (
    <TableCell colSpan={colSpan} className={classes.inputCell}>
      <InputBase classes={{ root: classes.input }} {...props} />
    </TableCell>
  );
};
const SelectCell = ({ children, ...props }) => {
  const classes = useStyles();
  return (
    <TableCell className={classes.inputCell}>
      <Select
        className={classes.selectInput}
        classes={{ select: classes.select }}
        MenuProps={{
          getContentAnchorEl: null,
          anchorOrigin: {
            vertical: "bottom",
            horizontal: "left"
          },
          classes: {
            paper: classes.selectMenu
          }
        }}
        {...props}
      >
        {children}
      </Select>
    </TableCell>
  );
};

const List = ({ total = false }) => {
  const classes = useStyles();
  const [rows, setRows] = React.useState([
    {
      buildingDescription: "30 Lake Road",
      internalArea: 200,
      inUse: "Yes",
      dateLastInUse: "12/08/1999"
    }
  ]);
  const [rowFocus, setRowFocus] = React.useState(null);

  const addRow = () => {
    setRows([
      ...rows,
      {
        buildingDescription: "45 River Way",
        internalArea: 470,
        inUse: "Yes",
        dateLastInUse: "12/08/2019"
      }
    ]);
  };
  const deleteRow = () => {
    const val = rows;
    val.pop();
    setRows([...val]);
  };
  return (
    <TableContainer>
      <Table className={classes.table}>
        <TableHead>
          <TableRow hover={false}>
            <TableCell className={classes.rowCount} />
            <TableCell>Building description</TableCell>
            <TableCell>
              Internal area in m<sup>2</sup>
            </TableCell>
            <TableCell>In use?</TableCell>
            <TableCell colSpan={2}>Date last in use?</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row, i) => {
            return (
              <TableRow key={i} className={rowFocus === i && classes.rowFocus}>
                <TableCell className={classes.rowCount}>{i + 1}</TableCell>
                <InputCell
                  placeholder="Add another"
                  value={row.buildingDescription}
                  onFocus={() => setRowFocus(i)}
                  onBlur={() => setRowFocus(null)}
                />
                <InputCell
                  placeholder="Internal area"
                  value={row.internalArea}
                  onFocus={() => setRowFocus(i)}
                  onBlur={() => setRowFocus(null)}
                  endAdornment={
                    <span>
                      m<sup>2</sup>
                    </span>
                  }
                />
                <SelectCell
                  value={row.inUse}
                  onFocus={() => setRowFocus(i)}
                  onBlur={() => setRowFocus(null)}
                >
                  <MenuItem value="Yes">Yes</MenuItem>
                  <MenuItem value="No">No</MenuItem>
                </SelectCell>
                <InputCell
                  onFocus={() => setRowFocus(i)}
                  onBlur={() => setRowFocus(null)}
                  placeholder="dd / mm / yyyy"
                  value={row.dateLastInUse}
                />
                <TableCell className={classes.deleteRow}>
                  <IconButton size="small" onClick={() => deleteRow()}>
                    <CloseIcon />
                  </IconButton>
                </TableCell>
              </TableRow>
            );
          })}
          <TableRow className={rowFocus === "new" && classes.rowFocus}>
            <TableCell className={classes.rowCount}>
              <Box color="#bbb">{rows.length + 1}</Box>
            </TableCell>
            <InputCell
              placeholder="Add another"
              onFocus={() => setRowFocus("new")}
              onBlur={() => setRowFocus(null)}
            />
            <InputCell
              placeholder="Internal area"
              onFocus={() => setRowFocus("new")}
              onBlur={() => setRowFocus(null)}
              endAdornment={
                <span>
                  m<sup>2</sup>
                </span>
              }
            />
            <SelectCell
              placeholder="Internal area"
              onFocus={() => setRowFocus("new")}
              onBlur={() => setRowFocus(null)}
            >
              <MenuItem value="y">Yes</MenuItem>
              <MenuItem value="n">No</MenuItem>
            </SelectCell>
            <InputCell
              colSpan={2}
              placeholder="dd / mm / yyyy"
              onFocus={() => setRowFocus("new")}
              onBlur={() => {
                setRowFocus(null);
                addRow();
              }}
            />
          </TableRow>
        </TableBody>
        {total && (
          <TableFooter>
            <TableRow hover={false}>
              <TableCell className={classes.rowCount} />
              <TableCell>
                <strong>Total</strong>
              </TableCell>
              <TableCell>
                200 m<sup>2</sup>
              </TableCell>
              <TableCell />
              <TableCell colSpan={2} />
            </TableRow>
          </TableFooter>
        )}
      </Table>
    </TableContainer>
  );
};

export default List;
