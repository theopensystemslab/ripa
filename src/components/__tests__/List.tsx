import { render } from "@testing-library/react";
import * as React from "react";

import List from "../List";

describe("List Component", () => {
  describe("without total bar", () => {
    it("should render snapshot", () => {
      const { asFragment } = render(<List />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("with total bar", () => {
    it("should render snapshot", () => {
      const { asFragment } = render(<List total />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
