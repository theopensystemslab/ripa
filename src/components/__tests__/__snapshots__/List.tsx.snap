// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component with total bar should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root makeStyles-table-111"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head makeStyles-rowCount-112"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Building description
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Internal area in m
            <sup>
              2
            </sup>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            In use?
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            colspan="2"
            scope="col"
          >
            Date last in use?
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-rowCount-112"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116"
            >
              <input
                class="MuiInputBase-input"
                placeholder="Add another"
                type="text"
                value="30 Lake Road"
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116 MuiInputBase-adornedEnd"
            >
              <input
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd"
                placeholder="Internal area"
                type="text"
                value="200"
              />
              <span>
                m
                <sup>
                  2
                </sup>
              </span>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectInput-117"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select makeStyles-select-118 MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                Yes
              </div>
              <input
                type="hidden"
                value="Yes"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116"
            >
              <input
                class="MuiInputBase-input"
                placeholder="dd / mm / yyyy"
                type="text"
                value="12/08/1999"
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-deleteRow-114"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-rowCount-112"
          >
            <div
              class="MuiBox-root MuiBox-root-210"
            >
              2
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116"
            >
              <input
                class="MuiInputBase-input"
                placeholder="Add another"
                type="text"
                value=""
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116 MuiInputBase-adornedEnd"
            >
              <input
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd"
                placeholder="Internal area"
                type="text"
                value=""
              />
              <span>
                m
                <sup>
                  2
                </sup>
              </span>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectInput-117"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select makeStyles-select-118 MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                <span>
                  â€‹
                </span>
              </div>
              <input
                placeholder="Internal area"
                type="hidden"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-115"
            colspan="2"
          >
            <div
              class="MuiInputBase-root makeStyles-input-116"
            >
              <input
                class="MuiInputBase-input"
                placeholder="dd / mm / yyyy"
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot
        class="MuiTableFooter-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-footer"
        >
          <td
            class="MuiTableCell-root MuiTableCell-footer makeStyles-rowCount-112"
          />
          <td
            class="MuiTableCell-root MuiTableCell-footer"
          >
            <strong>
              Total
            </strong>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-footer"
          >
            200 m
            <sup>
              2
            </sup>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-footer"
          />
          <td
            class="MuiTableCell-root MuiTableCell-footer"
            colspan="2"
          />
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;

exports[`List Component without total bar should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root makeStyles-table-1"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head makeStyles-rowCount-2"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Building description
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Internal area in m
            <sup>
              2
            </sup>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            In use?
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            colspan="2"
            scope="col"
          >
            Date last in use?
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-rowCount-2"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6"
            >
              <input
                class="MuiInputBase-input"
                placeholder="Add another"
                type="text"
                value="30 Lake Road"
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6 MuiInputBase-adornedEnd"
            >
              <input
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd"
                placeholder="Internal area"
                type="text"
                value="200"
              />
              <span>
                m
                <sup>
                  2
                </sup>
              </span>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectInput-7"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select makeStyles-select-8 MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                Yes
              </div>
              <input
                type="hidden"
                value="Yes"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6"
            >
              <input
                class="MuiInputBase-input"
                placeholder="dd / mm / yyyy"
                type="text"
                value="12/08/1999"
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-deleteRow-4"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-rowCount-2"
          >
            <div
              class="MuiBox-root MuiBox-root-100"
            >
              2
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6"
            >
              <input
                class="MuiInputBase-input"
                placeholder="Add another"
                type="text"
                value=""
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="1"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6 MuiInputBase-adornedEnd"
            >
              <input
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd"
                placeholder="Internal area"
                type="text"
                value=""
              />
              <span>
                m
                <sup>
                  2
                </sup>
              </span>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectInput-7"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select makeStyles-select-8 MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                <span>
                  â€‹
                </span>
              </div>
              <input
                placeholder="Internal area"
                type="hidden"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-inputCell-5"
            colspan="2"
          >
            <div
              class="MuiInputBase-root makeStyles-input-6"
            >
              <input
                class="MuiInputBase-input"
                placeholder="dd / mm / yyyy"
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
